lib/mongo/id.rb

```diff
@@ -0,0 +1,44 @@
+# Copyright (C) 2016-2019  MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Mongo
+  # This module abstracts the functionality for generating sequential unique integer IDs for
+  # instances of the class. It defines the method #next_id on the class that includes it. The
+  # implementation ensures that the IDs will be unique even when called from multiple threads.
+  #
+  # @example Define and use the Id module.
+  #   class Foo
+  #     include Id
+  #   end
+  #
+  #   f = Foo.new
+  #   foo.next_id # => 1
+  #   foo.next_id # => 2
+  #
+  # @since 2.7.0
+  module Id
+    def self.included(klass)
+      klass.class_variable_set(:@@id, 1)
```

Comment:
This assigns 1 to `id`, but then the next call to `next_id` also returns 1 which is confusing. Either `id` should be renamed to `@@next_id` to match the method, or the method should return the incremented number.

Reply:
The idea behind this is that `@@id` should not be used at all outside of this file, and classes should only use `next_id` to obtain the next ID which has not been used yet. The fact that this returns `1` first is a feature rather than a bug. I can add documentation that explains this.

Reply:
While the instance variables are indeed private, they should be used in a manner that is intuitive to understand. Prior to this PR ids started at 0 and the post-incremented value was returned. This behavior is one of two that I referred to in my previous comment.

Reply:
This isn't an instance variable, it's a class variable. If you feel that strongly about what the value of the class variable is even though it shouldn't ever be directly accessed, I'll change it.